File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.js
function solution(cap, n, deliveries, pickups) {
    
    const delivery = [];
    const pickup = [];
    
    for(let i=0; i<n; i++){
        delivery.push(deliveries[i]);
        pickup.push(pickups[i]);
    }
    
    var answer = 0;
    
    while(delivery.length || pickup.length){
        while(delivery.length && delivery[delivery.length-1] == 0){
            delivery.pop();
        }
        while(pickup.length && pickup[pickup.length-1] == 0){
            pickup.pop();
        }
        
        let max = Math.max(delivery.length, pickup.length);
        answer += max*2;
        
        let del = cap;
        let pic = cap;
        
        while(del > 0 && delivery.length){
            var pop = delivery.pop();
            if(pop > del){
                delivery.push(pop - del);
            }
            del -= pop;
        }
        
        while(pic > 0 && pickup.length){
            var pop = pickup.pop();
            if(pop > pic){
                pickup.push(pop - pic);
            }
            pic -= pop;
        }
        
    }
    

    return answer;
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.*;

class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        
        Deque<Integer> delivery = new ArrayDeque<>();
        Deque<Integer> pickup = new ArrayDeque<>();
        
        for(int i=0; i<n; i++){
            delivery.add(deliveries[i]);
            pickup.add(pickups[i]);
        }
        
        long answer = 0;
        
        while(!delivery.isEmpty() || !pickup.isEmpty()){
            int del = cap;
            int pic = cap;
            while(!delivery.isEmpty() && delivery.peekLast() == 0){
                delivery.pollLast();
            }
            while(!pickup.isEmpty() && pickup.peekLast() == 0){
                pickup.pollLast();
            }
            int len = Math.max(delivery.size(), pickup.size());
            answer += len*2;
            while(del > 0 && !delivery.isEmpty()){
                int poll = delivery.pollLast();
                if(poll > del){
                    delivery.add(poll-del);
                }
                del -= poll;
            }
            while(pic > 0 && !pickup.isEmpty()){
                int poll = pickup.pollLast();
                if(poll > pic){
                    pickup.add(poll-pic);
                }
                pic -= poll;
            }
        }
        
        return answer;
    }
}